# Palmer Penguins
***

This notebook contains my analysis of the palmer penguins data set. The palmer penguins data set contains data measurements for three different penguin species, the Chinstrap, Gentoo and Adélie penguin. The data was collected between 2007 - 2009 by Dr. Kristen Gorman with the [Palmer Station Long Term Ecological Research Program](https://lternet.edu/site/palmer-antarctica-lter/) on three different islands in the Palmer Archipelago, Antarctica. 

![The Palmer Penguins. Artwork by @allison_horst](https://sebastiancallh.github.io/ox-hugo/palmer-penguins.png)

The Palmer Penguins, artwork by @allison_horst






According to [AllisonHorst.github.io](https://github.com/allisonhorst/palmerpenguins/blob/main/README.md) "the goal of the palmerpenguins is to provide a great dataset for data exploration and visualization, as an alternative to iris". 

# TO DO

map of the islands if possible

variables section  - Diagram of the bill


Tidy up data types - currently a hot mess

Think about how to tidy up different variables I've assigned. Must be a cleaner and clearer way to do it.


Correlation part needs comments. 

Add research on the data set. Who uses it and why? Recent research?  Building on the work in literature (lol) and comparison to similiar work.

## Import Modules
***

Three different modules are required for this analysis:
* pandas
* numpy
* matplotlib.pyplot

Pandas is a python library for manipulating data and for performing data analysis. It can perform statistical calculations, find a correlations between two or more columns and it can be used to visualise data. The pandas package must first be loaded using the 'import' command to work with it. It is usually imported with the alias 'pd'. 

Numpy can perform a wide variety of mathematical calculations on arrays and matrices. All the elements in a numpy array should be of the same data type.

Matplotlib.pyplot is used to create plots

In [1]:
# Import pandas
import pandas as pd

# Import numpy
import numpy as np

# Import matplotlib.pylab
import matplotlib.pylab as plt

# Import seaborn
import seaborn as sns

## Importing the data set
***

The data set was imported as a csv file from [Seaborn Data](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv). The original raw data set is available on Alison Horst's [GitHub](https://github.com/allisonhorst/palmerpenguins/blob/main/inst/extdata/penguins_raw.csv). 

The data set used has the advantage in that it has been processed to remove extraneous information such as any NA in the data and the year that data was collected. The other minor difference is that male and female are capitalised in the Seaborn data set. 

In [2]:
# Load the penguins data set
df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv")

In [3]:
# Have a look. Check that the file has loaded correctly.
df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
3            NaN     NaN  
4         3450.0  FEMALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[344 rows x 7 columns]

## Overview of the data set and the variables it contains.
***

In [4]:
# df.head() prints out the first five rows of the csv file.
df.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE

In [5]:
# Counts the number of rows and columms in the data set.
df.shape

(344, 7)

In [6]:
# df.info() gives concise information about the dataframe.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [7]:
# A count of the number of missing variables.
df.isna().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

### Overiew of the data set

The data set has 344 rows and seven columns. Each row is an individual penguin and for each penguin there are seven possible variables.

The seven variables in the dataset are:
* species
* island
* bill_length in mm
* bill depth in mm
* flipper length in mm
* the body mass in g
* the sex of the penguin.


This artwork by Allison Horst nicely depicts the difference between the bill length and bill depth and how they were measured. In the original data set the word culmen was used in place of bill. 

![Bill length and bill depth](https://miro.medium.com/v2/resize:fit:640/format:webp/1*WSb-Y9xb3YSx9owW7_JuMw.png)

The Palmer Penguins, artwork by @allison_horst

There are no variables missing from two columns, the species and island columns. The other columns have some missing data. The sex column has the most numbers of missing variables with 11. Bill length, bill depth, flipper length and body mass all have two pieces of missing data.

## Data types for the data set.

When a data set is loaded pandas automatically assigns what it thinks is the best data type for the data. Numerical data, are categorised as integers or floats. Non-numerical data are automatically assigned as objects or strings. 

Pandas data type names have a slightly different name compared to their python data types name, as shown in the table below.

| Python | Pandas |
|--------|--------|
| string| object|
|float|float64|
|integer|int64|
|bool|bool|
| - |category|
|-|datetime64|

In [8]:
# Check the data type of the variables. 
df.dtypes

species               object
island                object
bill_length_mm       float64
bill_depth_mm        float64
flipper_length_mm    float64
body_mass_g          float64
sex                   object
dtype: object

The bill length, bill depth, flipper length and body mass were given the data type float64 by pandas. Species, island and sex are all non-numerical data and were assigned the object data type.

Bill length and bill depth were measured in mm with the result recorded to one decimal place. These are floating point numbers and pandas has correctly assigned them to the data type of float64.

After reviewing the raw data file I noticed that while flipper length and body mass were measured in mm they were not measured to the same degree of accuracy as the bill measurements and were recorded as whole numbers i.e. integers. The correct data type for these two columns should be int64.

After completing the data camp course, [working with categorical data](https://app.datacamp.com/learn/courses/working-with-categorical-data-in-python) I realised that I needed to investigate the species, island and sex columns and not to assume that the object data type that was assigned to them was correct. If a column contains a limited and fixed number of possible non-numerical values it is possible to assign it to the pandas data type of category. There is no corresponding data type in python. For large data sets changing the data type to category could potentially result in significant memory savings.

First I need to check that these columns fulfill the criteria to fit the category data type by finding out the number of unique entries in these columns.

In [9]:
# Unique names in the species column.
df["species"].unique()

array(['Adelie', 'Chinstrap', 'Gentoo'], dtype=object)

In [10]:
# Unique values in the island column.
df["island"].unique()

array(['Torgersen', 'Biscoe', 'Dream'], dtype=object)

In [11]:
# Unique names in the sex column.
df["sex"].unique()

array(['MALE', 'FEMALE', nan], dtype=object)


Species contains three different objects, which are the penguin species names, Adelie, Gentoo and Chinstrap. 
The names of the three islands are Biscoe, Dream and Torgersen. The sex of the penguins are either male or female with 11 missing values as stated previously.

I believe these three columns should be in the data type of category as they all have finite, limited values.

In [12]:
# Change the data type to category
df[['species', 'island', 'sex']] = df[['species', 'island', 'sex']].astype('category')


For consistancy I would like the formatting of the sex column to be in line with the formatting of the species and island columns i.e title case.

In [13]:
# Changing the format of the sex column
df['sex'] = df['sex'].str.title()
df.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female

When I attempted to changed the data type of the flipper and body mass columns using the astype method, I ran into difficulty as pandas was not happy with the two missing values in both the flipper length and body mass columns. 

Want to look at the rows with missing data. Luckily I can see from df.head() that rows 3 and rows 339 only have the island and species name recorded for those penguins. I will drop those two rows to see if i can then change the data type. Another way to deal with missing data is to fill the rows with the mean or median values but I have decided to simply drop them as 

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html

In [14]:
df.iloc[[3, 339]]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen             NaN            NaN                NaN   
339  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g  sex  
3            NaN  NaN  
339          NaN  NaN

In [19]:
df.dropna(thresh = 3, inplace = True)

In [20]:
df.isna().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  9
dtype: int64

In [25]:
df[['flipper_length_mm', 'body_mass_g']] = df[['flipper_length_mm', 'body_mass_g']].astype('int64')

In [26]:
df.dtypes

species              category
island               category
bill_length_mm        float64
bill_depth_mm         float64
flipper_length_mm       int64
body_mass_g             int64
sex                    object
dtype: object

In [ ]:
# Describe the data set. Summary statistics about the data file. 
df.describe()

In [ ]:
# Renaming and reorganising the dataset so that the species column is now the index column. This will make slicing using loc() easier.
penguins = df.set_index("species")
penguins.head()

In [ ]:
adelie = penguins.loc[['Adelie']]
adelie.describe()

In [ ]:
chinstrap = penguins.loc[['Chinstrap']]
chinstrap.describe()

In [ ]:
gentoo = penguins.loc[['Gentoo']]
gentoo.describe()

EMpty rows NaN

## Bar chart of the penguin population on each island.

A bar chart can be used to examine the penguin population on each island.

In [ ]:
# Unique variables in the island attribute
df["island"].unique()

In [ ]:
# Penguin population on each island.
df['island'].value_counts()

In [ ]:
# Plotting a bar chart of the penguin population on each island

# Setting the x-axis to the islands
xpoints = np.array(["Biscoe", "Dream", "Torgersen"])

# Setting the y-axis to the count of each penguin per island
ypoints = np.array(df['island'].value_counts())

plt.bar(xpoints, ypoints, edgecolor = "black")

plt.xlabel("Island")
plt.ylabel("No. of Penguins")
plt.title("No. of Penguins per Island")
plt.show()

### Investigating the penguin species that are present on each island. 

While the above bar chart is reasonable, it would be nice to get more information about the penguins and their relationship to the island. Some additional information that could be plotted to a bar chart is a breakdown of the penguin count by species on each island. 

To do this we need to know:
* the penguin species that are present on each island
* a count of the penguins by species for each island.

In [ ]:
# Penguin species present on each island.
df.groupby('island')['species'].unique()

In [ ]:
# A count of the different penguin species on each island.  
df.groupby('island')['species'].value_counts()

The results of the count of the different species on each island can be visualised using a bar chart. After much trial and error, a [Countplot in Seaborn](https://seaborn.pydata.org/generated/seaborn.countplot.html) appears to be the simplest method of making the required bar chart. https://www.geeksforgeeks.org/countplot-using-seaborn-in-python/?ref=ml_lbp

Firstly, seaborn must be imported. 

In [ ]:
# A breakdown of the penguin population on each island by species. I would like if this was stacked by colour if possible
sns.countplot(df, x = 'island',
            hue = 'species',
            edgecolor = 'black'
            )
plt.title('Penguin count by species on each island.')
plt.show()

__Summary of the penguin population on each island.__

The three islands are Biscoe, Dream and Torgersen. Biscoe is the most populous island with 168 penguins. Torgersen is the least populous island with only 52 penguins of the penguins studied. 

Both Biscoe and Dream have two penguin species, with only one penguin species present on Torgersen. Adelie penguins were found on all three islands, with Chinstrap and Gentoo found on Dream and Torgersen respectively.

## Histogram of an appropriate variable
***
Task: Make a histogram using an appropriate variable. 

My chosen variable is flipper lengths. WHY?

In [ ]:
# Plot of the distribution of flipper length (mm)
plt.hist(np.array(df["flipper_length_mm"]))
plt.xlabel("Flipper length (mm)")
plt.ylabel("No of observations")
plt.title("Variations in penguin flipper length")
plt.show()

The histogram would be expected to have a normal curve distribution but as we can see it doesn't and has a very unusual shape. Further analysis is required to explain the shape of the histogram.

__Normal Distribution__ADD reference

![Normal Distribution](https://upload.wikimedia.org/wikipedia/commons/thumb/4/44/Standard_Normal_Distribution.png/320px-Standard_Normal_Distribution.png)




As there are three different penguin species, I will now analyse the flipper length for each species.

In [ ]:
# Investigating the mean, standard deviation and median of the flipper length for the different penguin species.
df.groupby("species")["flipper_length_mm"].agg([np.mean, np.std, np.median, min, max])

In [ ]:
# Same as above but using pivot_table # Investigating the mean, standard deviation and median of the penguin species
df.pivot_table(values = "flipper_length_mm", index = "species", aggfunc=[np.mean, np.std, np.median])

The above table demonstrates that each of the penguin species have noticeably different flipper lengths. The Gentoo penguins have the largest flippers with the average flipper length of 217mm and a median value of 216mm. The Adelie penguins have the smallest flipper length of 189mm with a median of 190mm.

The bars on the right of the above histogram from approximately 208mm is the distribution of flipper lengths for the Gentoo penguin. The bars on the left are for the overlapping values of the Adelie and Chinstrap penguins.

An improved histogram is required to show the distribution of flipper lengths for the penguin species. 

In [ ]:
# Histogram of the variations in flipper length by species.

# This is my histogram for the task.

fig, ax = plt.subplots()
ax.hist(adelie['flipper_length_mm'], label = "Adelie", edgecolor = "black", alpha = 0.5)
ax.hist(chinstrap['flipper_length_mm'], label = "Chinstrap", edgecolor = "black", alpha = 0.5)
ax.hist(gentoo['flipper_length_mm'], label = "Gentoo", edgecolor = "black", alpha = 0.5)
ax.set_title("Variations in flipper length by species")
ax.set_xlabel("Flipper length (mm)")
ax.set_ylabel("No of observations")
ax.legend()
plt.show()

This histogram clearly demonstrates that each species has a distinct flipper size. It would be possible from flipper length to identify a penguin as a Gentoo penguin. It would be more difficult to classify the Adelie and the Chinstrap penguins by their flipper lengths alone as the values overlap.

# Project
***

Select two variables for the data set and provide an analysis of how correlated they are.

## Brainstorm

WHat two variables could I compare?

* Is flipper length or bill length/depth in proportion to the body mass
* Do heavier penguins have longer flippers or bill length?
* Bigger bills, bigger body mass? Bigger bills can eat more? 
* Compare male vs female within the chosen variables
* Some information on the sex of the penguin is missing for some penguins. Could I predict whether they are male or female based on the other traits. 
* Adelie penguins live on all three islands, are there any differences in flipper, bills, body mass? 

In [ ]:
df.groupby("species").agg({"flipper_length_mm": 'mean', 'bill_length_mm' : 'mean', 'bill_depth_mm' : 'mean', 'body_mass_g' : 'mean'})

In [ ]:
# Simple scatter plot to get started of flipper length (mm) vs body mass (g)
plt.scatter(penguins['flipper_length_mm'], penguins['body_mass_g'])
plt.title('Flipper length vs body mass')
plt.xlabel('Flipper length (mm)')
plt.ylabel('Body mass (g)')
plt.show()

## Scatter plot by species

The COMMENT ON PLOT



In [ ]:
fig, ax = plt.subplots()

ax.scatter(adelie['flipper_length_mm'], adelie['body_mass_g'], label = 'adelie')
ax.scatter(chinstrap['flipper_length_mm'], chinstrap['body_mass_g'], label = 'chinstrap')
ax.scatter(gentoo['flipper_length_mm'], gentoo['body_mass_g'], label = 'gentoo')
ax.set_xlabel('Flipper length (mm)')
ax.set_ylabel('Body masss (g)')
ax.set_title(' Flipper length (mm) vs Body mass (g)')
ax.legend()
plt.show()

HWat is correlation?

How correlated are the flipper length and body mass? 

In [ ]:
# Overall correlation between flipper length and body mass.
penguins['body_mass_g'].corr(penguins['flipper_length_mm'])

Calculating the correlation between body mass and flipper length by species

In [ ]:
# Correlation between flipper length and body mass for Adelie penguins.
adelie['body_mass_g'].corr(adelie['flipper_length_mm'])

In [ ]:
# Correlation between flipper length and body mass for Chinstrap penguins.
chinstrap['body_mass_g'].corr(chinstrap['flipper_length_mm'])

In [ ]:
# Correlation between flipper length and body mass for Gentoo penguins.
gentoo['body_mass_g'].corr(gentoo['flipper_length_mm'])

## Correlation between body mass and flipper length.

The results from a correlation are in the range of -1 to +1, with -1 indicating a strong negative correlation, 0 indicating no correlation and +1 is a strong positive correlationhttps://www.ncbi.[Common pitfalls in statistical analysis: The use of correlation techniques](nlm.nih.gov/pmc/articles/PMC5079093/).

When the overall correlation is calculated it appears that there is a strong poditive correlation between flipper length and body mass with a value of 0.87. However when the correlation is performed by penguin species that the true correlation is lower. Adelie penguins display a moderate correlation with a value of 0.47 and the Gentoo penguins have a 

element wise correlation https://www.geeksforgeeks.org/how-to-calculate-correlation-between-two-columns-in-pandas/

In [ ]:
# Calculating the correlation between all the variables in the Adelie penguin. 
adelie_corr = adelie.drop(['island', 'sex'], axis = 1).corr()
adelie_corr

Overall for the Adelie penguin there is a weak to moderate correlation between the different variables. The strongest correlation in the table is between bill depth and body mass at 0.58, with bill length and body mass closely behind it at 0.55. 

While we can read the above table, it is not very visual and it has to be carefully analysed. [Seaborn heatmaps](https://www.datacamp.com/tutorial/seaborn-heatmaps) are very useful tool to visualise numerical data. The data is organised in a grid by colour, with the colour indicating order of magnitude. It is much easier to identify relationships between the variables of interest. 

As heatmaps require numerical data, I dropped the columns for island and sex for all species data sets. To make comparsions between the species easier to identify, I adjusted the scale of the heatmap so that the same colour corresponded to the same order of magnitude. 

In [ ]:
sns.heatmap(adelie.drop(['island', 'sex'], axis = 1).corr(), annot = True, linewidths = 0.2, vmin = 0.2, vmax = 1)
plt.title('Heatmap for Adelie Penguins', y= 1.02)
plt.show()

In [ ]:
sns.heatmap(chinstrap.drop(['island','sex'], axis = 1).corr(), annot = True, linewidths = 0.2, vmin = 0.2, vmax = 1)
plt.title('Heatmap for Chinstrap Penguins', y= 1.02)
plt.show()

In [ ]:
sns.heatmap(gentoo.drop(['island','sex'], axis = 1).corr(), annot = True, linewidths = 0.2, vmin = 0.2, vmax = 1)
plt.title('Heatmap for Gentoo Penguins', y= 1.02)
plt.show()

While a straight comparison appears to be a good correlation but when it is check by species. Poor correlation between body mass and flipper length for adelie penguins. Gentoo has a moderat( check correct term) correlation. 

Study of male vs female within the three different species. How would i plot it?

scatter plot by species male vs female
plot for each of 4 variables. 

In [ ]:
male_adelie = adelie[(adelie['sex'] ==  'Male')]
female_adelie = adelie[(adelie['sex'] ==  'Female')]

male_chinstrap = chinstrap[(chinstrap['sex']) == 'Male']
female_chinstrap = chinstrap[(chinstrap['sex']) == 'Female']

In [ ]:
fig, ax = plt.subplots()

ax.scatter(male_adelie['flipper_length_mm'], male_adelie['body_mass_g'], label = 'Male') 
ax.scatter(female_adelie['flipper_length_mm'], female_adelie['body_mass_g'], label = 'Female')
ax.set_ylabel('Body mass (g)')
ax.set_title('Flipper length vs body mass for male and female adelie penguins')
ax.legend()
plt.show()


In [ ]:
male_adelie['body_mass_g'].corr(male_adelie['flipper_length_mm'])

In [ ]:
female_adelie['body_mass_g'].corr(female_adelie['flipper_length_mm'])

In [ ]:
# Comparison of flipper lengths and body mass by species.
g = sns.relplot(x = 'flipper_length_mm', y = 'body_mass_g', hue = 'species', col = 'sex', data = penguins)
g.fig.suptitle('Comparison of flipper lengths vs body mass by sex', y = 1.05)
g.set(xlabel = 'flipper length (mm)', ylabel = 'Body mass (g)')


plt.show()

In [ ]:
#import warnings
#warnings.filterwarnings('ignore')

g = sns.relplot(x = 'flipper_length_mm', y = 'body_mass_g', col = 'species', hue = 'sex', data = penguins)
g.fig.suptitle('Comparison of flipper lengths vs body mass by species and sex', y = 1.05)
g.set(xlabel = 'Flipper length (mm)', ylabel = 'Body mass (g)')
plt.show()

Any differences between the Adelie penguins on the three different islands?

## Missing data 

What data is missing in the adelie df?
Any duplicates?


In [ ]:
adelie.isnull().sum()


In [ ]:
adelie.duplicated().sum()

In [ ]:
adelie_no_na = adelie.dropna()

In [ ]:
adelie_no_na['body_mass_g'].corr(adelie_no_na['flipper_length_mm'])

In [ ]:
adelie_no_na.groupby('island')['body_mass_g'].mean().std()

In [ ]:
adelie.groupby('island')['body_mass_g'].mean().std()

** Investigating the Penguin Species.

In [ ]:
df.groupby('species').count()

Gentoo penguins, missing sex data. Can I tell if guess(?, need better word, if they are male or female.)

1st get the missing rows.


In [ ]:
gentoo_missing = gentoo[~gentoo['sex'].isin(['MALE', 'FEMALE'])]
gentoo_missing


In [ ]:
df.groupby('island')['body_mass_g'].mean()


In [ ]:
df.groupby('species')['body_mass_g'].mean()

In [ ]:
# Number and sex of penguins
df["sex"].value_counts()

In [ ]:
# Are male penguins heavier than female penguins?

df.groupby('sex')['body_mass_g'].mean()

In [ ]:
df.groupby('species')['flipper_length_mm'].mean()


In [ ]:
# Same as above but using pivot_table

df.pivot_table(values = "flipper_length_mm", index = "species", aggfunc=[np.mean, np.std, np.median])

### References

Datacamp [Introduction to Data Visualization with Matplotlib](https://app.datacamp.com/learn/courses/introduction-to-data-visualization-with-matplotlib)

Datacamp [Data Manipulation with Pandas](https://app.datacamp.com/learn/courses/data-manipulation-with-pandas)

Datacamp [Introduction to Data Visualization with Seaborn](https://app.datacamp.com/learn/courses/introduction-to-data-visualization-with-seaborn)


## End

***